lib/services/database_test.go

```diff
@@ -155,40 +156,135 @@ func TestDatabaseFromRDSCluster(t *testing.T) {
+		require.NoError(t, err)
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("reader", func(t *testing.T) {
+		expected, err := types.NewDatabaseV3(types.Metadata{
+			Name:        "cluster-1-reader",
+			Description: "Aurora cluster in us-east-1 (reader endpoint)",
+			Labels: map[string]string{
+				types.OriginLabel:  types.OriginCloud,
+				labelAccountID:     "1234567890",
+				labelRegion:        "us-east-1",
+				labelEngine:        RDSEngineAuroraMySQL,
+				labelEngineVersion: "8.0.0",
+				labelEndpointType:  "reader",
+				"key":              "val",
+			},
+		}, types.DatabaseSpecV3{
+			Protocol: defaults.ProtocolMySQL,
+			URI:      "reader.host:3306",
+			AWS:      expectedAWS,
+		})
+		require.NoError(t, err)
+		actual, err := NewDatabaseFromRDSClusterReaderEndpoint(cluster)
+		require.NoError(t, err)
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("custom endpoints", func(t *testing.T) {
+		expectedLabels := map[string]string{
 			types.OriginLabel:  types.OriginCloud,
 			labelAccountID:     "1234567890",
 			labelRegion:        "us-east-1",
 			labelEngine:        RDSEngineAuroraMySQL,
 			labelEngineVersion: "8.0.0",
+			labelEndpointType:  "custom",
 			"key":              "val",
-		},
-	}, types.DatabaseSpecV3{
-		Protocol: defaults.ProtocolMySQL,
-		URI:      "localhost:3306",
-		AWS: types.AWS{
-			AccountID: "1234567890",
-			Region:    "us-east-1",
-			RDS: types.RDS{
-				ClusterID:  "cluster-1",
-				ResourceID: "resource-1",
-				IAMAuth:    true,
+		}
+
+		expectedMyEndpoint1, err := types.NewDatabaseV3(types.Metadata{
+			Name:        "cluster-1-custom-myendpoint1",
+			Description: "Aurora cluster in us-east-1 (custom endpoint)",
+			Labels:      expectedLabels,
+		}, types.DatabaseSpecV3{
+			Protocol: defaults.ProtocolMySQL,
+			URI:      "myendpoint1.cluster-custom-example.us-east-1.rds.amazonaws.com:3306",
+			AWS:      expectedAWS,
+			TLS: types.DatabaseTLS{
+				ServerName: "localhost",
 			},
-		},
+		})
+		require.NoError(t, err)
+
+		expectedMyEndpoint2, err := types.NewDatabaseV3(types.Metadata{
+			Name:        "cluster-1-custom-myendpoint2",
+			Description: "Aurora cluster in us-east-1 (custom endpoint)",
+			Labels:      expectedLabels,
+		}, types.DatabaseSpecV3{
+			Protocol: defaults.ProtocolMySQL,
+			URI:      "myendpoint2.cluster-custom-example.us-east-1.rds.amazonaws.com:3306",
+			AWS:      expectedAWS,
+			TLS: types.DatabaseTLS{
+				ServerName: "localhost",
+			},
+		})
+		require.NoError(t, err)
+
+		databases, err := NewDatabasesFromRDSClusterCustomEndpoints(cluster)
+		require.NoError(t, err)
+		require.Equal(t, types.Databases{expectedMyEndpoint1, expectedMyEndpoint2}, databases)
 	})
+
+	t.Run("bad custom endpoints ", func(t *testing.T) {
+		badCluster := *cluster
+		badCluster.CustomEndpoints = []*string{
+			aws.String("badendpoint1"),
+			aws.String("badendpoint2"),
+		}
+		_, err := NewDatabasesFromRDSClusterCustomEndpoints(&badCluster)
+		require.Error(t, err)
```

Comment:
Can we check what error is being returned here? Function maybe failing for a different reason than we expect.

Reply:
the errors are aggregated so we can go through as many endpoints as we can. thus it's hard to valid against the aggregated error here.